code/game/atoms_movable.dm

```diff
@@ -518,3 +517,39 @@
 	var/turf/currentturf = get_turf(src)
 	if(currentturf && (currentturf.z == ZLEVEL_CENTCOM || currentturf.z == ZLEVEL_STATION))
 		. = TRUE
+
+
+/* Language procs */
+/atom/movable/proc/grant_language(datum/language/dt)
+	if(!has_language(dt))
+		var/datum/language/L = new dt
+		LAZYADD(languages, L)
+
+/atom/movable/proc/grant_all_languages(ignore_restrictions=FALSE)
+	for(var/la in subtypesof(/datum/language))
+		grant_language(la)
+
+	if(ignore_restrictions)
+		SET_SECONDARY_FLAG(src, CAN_ALWAYS_SPEAK_A_LANGUAGE)
+
+/atom/movable/proc/remove_language(datum/language/dt)
+	if(languages && languages.len)
+		for(var/L in languages)
+			if(istype(L, dt))
+				languages -= L
+				qdel(L)
+				break
+
+/atom/movable/proc/has_language(datum/language/dt)
+	. = FALSE
+	if(languages)
+		for(var/L in languages)

```

Comment:
@coiax a typecache is just an assoc list.

```
var/list/L = list(/type/a, /type/b, /type/c)
var/list/T = typecacheof(L)
->
var/list/T = list(
    /type/a = TRUE,
    /type/b = TRUE,
    /type/c = TRUE
)
```

This means, obviously, that you can modify it as you would any other assoc list.